packages/compass-workspaces/src/stores/workspaces.ts

```diff
@@ -843,28 +858,71 @@ export const openTabFromCurrent = (
   };
 };
 
-type CloseTabAction = { type: WorkspacesActions.CloseTab; atIndex: number };
+type DuplicateTabAction = {
+  type: WorkspacesActions.DuplicateTab;
+  atIndex: number;
+};
+
+export const duplicateTab = (atIndex: number): DuplicateTabAction => {
+  return {
+    type: WorkspacesActions.DuplicateTab,
+    atIndex,
+  };
+};
+
+async function confirmClosingTab() {
+  return await showConfirmation({
+    title: 'Are you sure you want to close the tab?',
+    description:
+      'The content of this tab has been modified. You will lose your changes if you close it.',
+    buttonText: 'Close tab',
+    variant: 'danger',
+    'data-testid': 'confirm-tab-close',
+  });
+}
+
+type CloseTabsAction = { type: WorkspacesActions.CloseTabs; tabIds: string[] };
 
 export const closeTab = (
   atIndex: number
-): WorkspacesThunkAction<Promise<void>, CloseTabAction> => {
+): WorkspacesThunkAction<Promise<void>, CloseTabsAction> => {
   return async (dispatch, getState) => {
-    const tab = getState().tabs[atIndex];
-    if (!canCloseTab(tab)) {
-      const confirmClose = await showConfirmation({
-        title: 'Are you sure you want to close the tab?',
-        description:
-          'The content of this tab has been modified. You will lose your changes if you close it.',
-        buttonText: 'Close tab',
-        variant: 'danger',
-        'data-testid': 'confirm-tab-close',
-      });
-      if (!confirmClose) {
-        return;
-      }
+    const { tabs } = getState();
+    const tab = tabs[atIndex];
+    if (canCloseTab(tab) || (await confirmClosingTab())) {
+      dispatch({ type: WorkspacesActions.CloseTabs, tabIds: [tab.id] });
+      cleanupRemovedTabs(tabs, getState().tabs);
     }
-    dispatch({ type: WorkspacesActions.CloseTab, atIndex });
-    cleanupLocalAppRegistryForTab(tab?.id);
+  };
+};
+
+export const closeAllOtherTabs = (
+  atIndex: number
+): WorkspacesThunkAction<Promise<void>, CloseTabsAction | SelectTabAction> => {
+  return async (dispatch, getState) => {
+    const { tabs } = getState();
+    const tabsToClose = await tabs.reduce(
+      async (prev: Promise<WorkspaceTab[]>, tab, tabIndex) => {
+        const tabsToClose = await prev;
+        if (tabIndex === atIndex) {
+          return tabsToClose; // Skip the tab which is not being closed
+        }
+        if (!canCloseTab(tab)) {
+          // Select the closing tab - to show the confirmation dialog in context
+          dispatch({ type: WorkspacesActions.SelectTab, atIndex: tabIndex });
+          if (!(await confirmClosingTab())) {
+            return tabsToClose; // Skip this tab
+          }
+        }
+        return [...tabsToClose, tab];
+      },
+      Promise.resolve([])
+    );
```

Comment:
Nit: I found it really hard to read through this reducer with the promise being an accumulator that always needs to be awaited, I think a for of loop reads easier, but feel free to ignore this suggestion

```suggestion
    const tabsToClose = [];
    for (const [tabIndex, tab] of tabs.entries()) {
      if (tabIndex !== atIndex && !canCloseTab(tab)) {
        // Select the closing tab - to show the confirmation dialog in context
        dispatch({ type: WorkspacesActions.SelectTab, atIndex: tabIndex });
        if (await confirmClosingTab()) {
          tabsToClose.push(tab);
        }
      }
    }
```

Reply:
You're right - a simple loop reads much better. Thanks!

Reply:
I've pushed an update using a for loop. I needed to change it slightly to accommodate for tabs to which `canCloseTab` returned true.

Reply:
Yeah, sorry, I sketched the code, but forgot to double-check it when making a suggestion, thanks for refactoring, looks good to me!